File: src/main/java/com/zepben/vertxutils/routing/RouteVersionUtils.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.routing;

import com.zepben.annotations.EverythingIsNonnullByDefault;

import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@EverythingIsNonnullByDefault
@SuppressWarnings("WeakerAccess")
public class RouteVersionUtils {

    public static <T extends VersionableRoute> List<Route> forVersion(T[] availableRoutes, int version, Function<T, Route> routeFactory) {
        return Stream.of(availableRoutes)
            .filter(rv -> rv.routeVersion().includes(version))
            .map(routeFactory)
            .collect(Collectors.toList());
    }

}

File: src/main/java/com/zepben/vertxutils/routing/RouteVersionUtils.kt
/*
 * Copyright 2026 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.routing

object RouteVersionUtils {

    inline fun <reified T> ((T) -> Route?).forVersion(version: Int): List<Route> where T : Enum<T>, T : VersionableRoute =
        enumValues<T>()
            .asSequence()
            .filter { version in it.routeVersion }
            .mapNotNull { this(it) }
            .toList()

}
